airflow/jobs.py

```diff
@@ -1064,11 +1064,16 @@ def _execute_task_instances(self,
                 dag_id = task_instance.dag_id
 
                 if dag_id not in dag_id_to_possibly_running_task_count:
+                    # TODO(saguziel): we should also count QUEUED in the below check
+                    # right now, there can be failed dag runs with queued tasks
+                    # due to an unknown race condition
+                    # this should still respect concurrency given that there is a worker
+                    # side check for concurrency
```

Comment:
A little bit hard to read, how about:

Tasks with the QUEUED state should also be counted below, but for now we cannot count them. This is because there is no guarantee that queued tasks in failed dagruns will or will not eventually run and queued tasks that will never run will consume slots and can stall a DAG. Once we can guarantee that all queued tasks in failed dagruns will never run (e.g. make sure that all running/newly queued TIs have running dagruns), then we can include QUEUED tasks here, with the constraint that they are in running dagruns.